esp-bootloader-esp-idf/src/partitions.rs

```diff
@@ -0,0 +1,765 @@
+    {
+        FlashRegion { raw: self, flash }
+    }
+}
+
+impl core::fmt::Debug for PartitionEntry<'_> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("PartitionEntry")
+            .field("magic", &self.magic())
+            .field("raw_type", &self.raw_type())
+            .field("raw_subtype", &self.raw_subtype())
+            .field("offset", &self.offset())
+            .field("len", &self.len())
+            .field("label", &self.label_as_str())
+            .field("flags", &self.flags())
+            .field("is_read_only", &self.is_read_only())
+            .field("is_encrypted", &self.is_encrypted())
+            .finish()
+    }
+}
+
+#[cfg(feature = "defmt")]
+impl defmt::Format for PartitionEntry<'_> {
+    fn format(&self, fmt: defmt::Formatter) {
+        defmt::write!(
+            fmt,
+            "PartitionEntry (\
+            magic = {}, \
+            raw_type = {}, \
+            raw_subtype = {}, \
+            offset = {}, \
+            len = {}, \
+            label = {}, \
+            flags = {}, \
+            is_read_only = {}, \
+            is_encrypted = {}\
+            )",
+            self.magic(),
+            self.raw_type(),
+            self.raw_subtype(),
+            self.offset(),
+            self.len(),
+            self.label_as_str(),
+            self.flags(),
+            self.is_read_only(),
+            self.is_encrypted()
+        )
+    }
+}
+
+/// Errors which can be returned.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, strum::Display)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// The partition table is invalid.
+    Invalid,
+    /// An operation tries to access data that is out of bounds.
+    OutOfBounds,
+    /// An error which originates from the embedded-storage implementation.
+    StorageError,
+    /// The partition is write protected.
+    WriteProtected,
+}
+
+impl core::error::Error for Error {}
+
+/// A partition table.
+pub struct PartitionTable<'a> {
+    binary: &'a [[u8; RAW_ENTRY_LEN]],
+    entries: usize,
+}
+
+impl<'a> PartitionTable<'a> {
+    fn new(binary: &'a [u8]) -> Result<Self, Error> {
+        if binary.len() > PARTITION_TABLE_MAX_LEN {
+            return Err(Error::Invalid);
+        }
+
+        if binary.len() % RAW_ENTRY_LEN != 0 {
+            return Err(Error::Invalid);
+        }
+
+        // we checked binary before
+        let binary = unsafe {
+            core::slice::from_raw_parts(
+                binary.as_ptr() as *const [u8; RAW_ENTRY_LEN],
+                binary.len() / RAW_ENTRY_LEN,
+            )
+        };
+
+        let mut raw = Self {
+            binary,
+            entries: binary.len() / RAW_ENTRY_LEN,
+        };
+
+        #[cfg(feature = "validation")]
+        {
+            let (hash, index) = {
+                let mut i = 0;
+                loop {
+                    let entry = raw.get_partition(i).unwrap();
```

Comment:
This should be `unwrap!`, I'm not so keen on including panic strings that may not even be possible to be hit.

Reply:
I hesitated to bring `fmt.rs` yet - but why not - sooner or later we want to actually log something

Reply:
Since I question unwrapping here anyway, let's not be hasty about that :)

Reply:
I guess bringing some macros is fine - but I just realized the unit tests don't run in CI and I broke it at some point between opening the PR and now 😢

Reply:
ok - we don't run any on-host tests in CI (and e.g. esp-config includes some) ... not good